### PR TITLE
Make consistent store watches cancellable

### PR DESCRIPTION
**Pull request by AlCutter**

### CODE REVIEW

File: cpp/log/cluster_state_controller.h [hunk 1]
```diff
@@ -65,10 +66,12 @@ class ClusterStateController {
   // Thread entry point for ServingSTH updater thread.
   void ClusterServingSTHUpdater();
 
+  util::Executor* const executor_;        // Not owned by us
```
Comment (pphaneuf) [inline cpp/log/cluster_state_controller.h]: You're not using this, might as well leave it to ```watch_node_states_task_``` to keep it.

File: cpp/log/etcd_consistent_store-inl.h [hunk 2]
```diff
@@ -277,6 +265,61 @@ util::Status EtcdConsistentStore<Logged>::SetClusterNodeState(
 
 
 template <class Logged>
+template <class T, class CB>
+void EtcdConsistentStore<Logged>::ConvertSingleUpdate(
```
Comment (pphaneuf) [inline cpp/log/etcd_consistent_store-inl.h]: I don't think this needs to be a member.

File: cpp/log/etcd_consistent_store-inl.h [hunk 3]
```diff
@@ -277,6 +265,61 @@ util::Status EtcdConsistentStore<Logged>::SetClusterNodeState(
...
+    callback(Update<T>(EntryHandle<T>(), false /* exists */));
+  } else {
+    callback(TypedUpdateFromWatchUpdate<T>(updates[0]));
+  }
+}
+
+
+template <class Logged>
+template <class T, class CB>
+void EtcdConsistentStore<Logged>::ConvertMultipleUpdate(
```
Comment (pphaneuf) [inline cpp/log/etcd_consistent_store-inl.h]: Same here, this should just be a helper free function.

File: cpp/log/etcd_consistent_store.h [hunk 4]
```diff
@@ -86,29 +95,40 @@ class EtcdConsistentStore : public ConsistentStore<Logged> {
 
   std::string GetFullPath(const std::string& key) const;
 
-  // Static member just so that it has friend access to the private
-  // c'tor/setters of Update<>
+
+  // The following 3 methods are static just so that they have friend access to
+  // the private c'tor/setters of Update<>
```
Comment (pphaneuf) [inline cpp/log/etcd_consistent_store.h]: Bleh, nevermind... Maybe add a ```/* static */``` comment to the two implementations?

File: cpp/log/etcd_consistent_store.h [hunk 5]
```diff
@@ -86,29 +95,40 @@ class EtcdConsistentStore : public ConsistentStore<Logged> {
...
+  static void ConvertMultipleUpdate(
+      const CB& callback, const std::vector<EtcdClient::WatchUpdate>& updates);
+
+  // Converts a generic WatchUpdate to an Update<T>.
+  // T must implement ParseFromString().
   template <class T>
   static Update<T> TypedUpdateFromWatchUpdate(
       const EtcdClient::WatchUpdate& update);
 
+
```
Comment (pphaneuf) [inline cpp/log/etcd_consistent_store.h]: Spurious newline.

File: cpp/log/cluster_state_controller.h (hunk 1 above)
Comment (AlCutter) [inline cpp/log/cluster_state_controller.h]: dinged.

**Comment by pphaneuf** [comment]
Done with the review, just a few tiny details...

File: cpp/log/etcd_consistent_store-inl.h (hunk 2 above)
Comment (AlCutter) [inline cpp/log/etcd_consistent_store-inl.h]: It needs `TypedUpdateFromWatchUpdate()`...

File: cpp/log/etcd_consistent_store-inl.h (hunk 3 above)
Comment (AlCutter) [inline cpp/log/etcd_consistent_store-inl.h]: same here

File: cpp/log/etcd_consistent_store.h (hunk 4 above)
Comment (AlCutter) [inline cpp/log/etcd_consistent_store.h]: Added

File: cpp/log/etcd_consistent_store.h (hunk 5 above)
Comment (AlCutter) [inline cpp/log/etcd_consistent_store.h]: zapped

**Comment by pphaneuf** [comment]
LGTM

**Pull request merged**